src/dataflow-types/src/client/controller/storage.rs

```diff
@@ -386,18 +393,39 @@ where
         &mut self,
         commands: Vec<(GlobalId, Vec<Update<Self::Timestamp>>, Self::Timestamp)>,
     ) -> Result<(), StorageError> {
-        for (id, updates, upper) in commands.iter() {
-            for update in updates {
-                if !update.timestamp.less_than(upper) {
-                    return Err(StorageError::UpdateBeyondUpper(*id));
+        for (id, updates, new_upper) in commands {
+            for update in &updates {
+                if !update.timestamp.less_than(&new_upper) {
+                    return Err(StorageError::UpdateBeyondUpper(id));
                 }
             }
+            let upper = self.collection(id)?.write_frontier.frontier().to_owned();
+            let new_upper = Antichain::from_elem(new_upper);
+
+            let handles = self
+                .state
+                .persist_handles
+                .get_mut(&id)
+                .expect("unknown collection id");
+
+            let updates = updates
+                .into_iter()
+                .map(|u| (((), SourceData(Ok(u.row))), u.timestamp, u.diff));
+
+            handles
+                .write
+                .append(updates, upper.clone(), new_upper.clone())
+                .await
+                .expect("cannot append updates")
+                .expect("cannot append updates")
+                .expect("invalid/outdated upper");
```

Comment:
why is it okay to panic on this error? can't we transiently fail here, and need to retry?

Reply:
We should transiently fail, yes. The caller should figure out what to do with their append, it can't be simply retried

Reply:
why do we panic then?